`PermissionDenied` panic when writing `sdylib` into nonexisting directory

👍 
```rust
#![crate_type = "lib"]
fn main() {}
```
`rustc a.rs -o/dev/null`
=>
```
warning: function `main` is never used
 --> a.rs:2:4
  |
2 | fn main() {}
  |    ^^^^
  |
  = note: `#[warn(dead_code)]` on by default

error: couldn't create a temp dir: Permission denied (os error 13) at path "/dev/rmetaIawTrc"

error: aborting due to 1 previous error; 1 warning emitted
```

👎 
```rust
#![crate_type = "sdylib"]
fn main() {}
```
`rustc a.rs -o/dev/null`

```
error[E0658]: `sdylib` crate type is unstable
  |
  = note: see issue #139939 <https://github.com/rust-lang/rust/issues/139939> for more information
  = help: add `#![feature(export_stable)]` to the crate attributes to enable
  = note: this compiler was built on 2025-07-14; consider upgrading it if it is out of date


thread 'rustc' panicked at compiler/rustc_interface/src/passes.rs:863:61:
called `Result::unwrap()` on an `Err` value: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
stack backtrace:
   0:     0x73c7dfbe9343 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h4e562cc181463c4a
   1:     0x73c7e04027b7 - core::fmt::write::h5e7353c331f38806
   2:     0x73c7dfbdee73 - std::io::Write::write_fmt::h717a595a455455fa
   3:     0x73c7dfbe91a2 - std::sys::backtrace::BacktraceLock::print::heda4454f705b9f66
   4:     0x73c7dfbecc47 - std::panicking::default_hook::{{closure}}::he59e210b9c61086c
   5:     0x73c7dfbec7ab - std::panicking::default_hook::h906ec2af00bcd37b
   6:     0x73c7dec85333 - std[bd2dd26c0c142468]::panicking::update_hook::<alloc[ceda81b4d5bba64f]::boxed::Box<rustc_driver_impl[5491146aeef0441]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x73c7dfbed48e - std::panicking::rust_panic_with_hook::h0fd2f96c6f3d7b64
   8:     0x73c7dfbed18a - std::panicking::begin_panic_handler::{{closure}}::h852fd4cdebbb1bdd
   9:     0x73c7dfbe9809 - std::sys::backtrace::__rust_end_short_backtrace::hace051c048bd1b6f
  10:     0x73c7dfbece6d - __rustc[73a2578ac8587ebb]::rust_begin_unwind
  11:     0x73c7dc2ed850 - core::panicking::panic_fmt::h56753bffe2cf3db9
  12:     0x73c7dc36ff56 - core::result::unwrap_failed::h02609c1e70c431ee
  13:     0x73c7e17233ac - rustc_interface[bf609510522f04e0]::passes::write_interface
  14:     0x73c7e1723ca6 - rustc_interface[bf609510522f04e0]::passes::create_and_enter_global_ctxt::<core[5857717783b92e2b]::option::Option<rustc_interface[bf609510522f04e0]::queries::Linker>, rustc_driver_impl[5491146aeef0441]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}
  15:     0x73c7e1711c14 - rustc_interface[bf609510522f04e0]::interface::run_compiler::<(), rustc_driver_impl[5491146aeef0441]::run_compiler::{closure#0}>::{closure#1}
  16:     0x73c7e15d04b8 - std[bd2dd26c0c142468]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[bf609510522f04e0]::util::run_in_thread_with_globals<rustc_interface[bf609510522f04e0]::util::run_in_thread_pool_with_globals<rustc_interface[bf609510522f04e0]::interface::run_compiler<(), rustc_driver_impl[5491146aeef0441]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  17:     0x73c7e15d0196 - <<std[bd2dd26c0c142468]::thread::Builder>::spawn_unchecked_<rustc_interface[bf609510522f04e0]::util::run_in_thread_with_globals<rustc_interface[bf609510522f04e0]::util::run_in_thread_pool_with_globals<rustc_interface[bf609510522f04e0]::interface::run_compiler<(), rustc_driver_impl[5491146aeef0441]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[5857717783b92e2b]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  18:     0x73c7e15d60fd - std::sys::pal::unix::thread::Thread::new::thread_start::hbf083949671eb05c
  19:     0x73c7daea07eb - <unknown>
  20:     0x73c7daf2418c - <unknown>
  21:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/home/matthias/vcs/github/rust_misc_stuff/tests/ui/ATTR/o/rustc-ice-2025-07-15T16_50_59-1567994.txt` to your bug report

query stack during panic:
end of query stack
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0658`.
```

```
rustc 1.90.0-nightly (a00149764 2025-07-14)
binary: rustc
commit-hash: a001497644bc229f1abcc5b2528733386591647f
commit-date: 2025-07-14
host: x86_64-unknown-linux-gnu
release: 1.90.0-nightly
LLVM version: 20.1.8
```

--- a/compiler/rustc_interface/src/diagnostics.rs
+++ b/compiler/rustc_interface/src/diagnostics.rs
@@ -84,7 +84,7 @@ pub(crate) struct TempsDirError;
 pub(crate) struct OutDirError;
 
 #[derive(Diagnostic)]
-#[diag("failed to write file {$path}: {$error}\"")]
+#[diag("failed to write file {$path}: {$error}")]
 pub(crate) struct FailedWritingFile<'a> {
     pub path: &'a Path,
     pub error: io::Error,

--- a/compiler/rustc_interface/src/passes.rs
+++ b/compiler/rustc_interface/src/passes.rs
@@ -881,9 +881,11 @@ pub fn write_interface<'tcx>(tcx: TyCtxt<'tcx>) {
         &tcx.sess.psess.attr_id_generator,
     );
     let export_output = tcx.output_filenames(()).interface_path();
-    let mut file = fs::File::create_buffered(export_output).unwrap();
-    if let Err(err) = write!(file, "{}", krate) {
-        tcx.dcx().fatal(format!("error writing interface file: {}", err));
+    let mut file = fs::File::create_buffered(&export_output).unwrap_or_else(|error| {
+        tcx.dcx().emit_fatal(diagnostics::FailedWritingFile { path: &export_output, error })
+    });
+    if let Err(error) = write!(file, "{}", krate) {
+        tcx.dcx().emit_fatal(diagnostics::FailedWritingFile { path: &export_output, error });
     }
 }
 

--- a/compiler/rustc_middle/src/error.rs
+++ b/compiler/rustc_middle/src/error.rs
@@ -18,7 +18,7 @@ pub(crate) struct DropCheckOverflow<'tcx> {
 }
 
 #[derive(Diagnostic)]
-#[diag("failed to write file {$path}: {$error}\"")]
+#[diag("failed to write file {$path}: {$error}")]
 pub(crate) struct FailedWritingFile<'a> {
     pub path: &'a Path,
     pub error: io::Error,

--- a/tests/run-make/export/write-interface-nonexistent-dir/libr.rs
+++ b/tests/run-make/export/write-interface-nonexistent-dir/libr.rs
@@ -1,0 +1,13 @@
+#![feature(export_stable)]
+#![crate_type = "sdylib"]
+
+pub mod m {
+    #[repr(C)]
+    pub struct S {
+        pub x: i32,
+    }
+
+    pub extern "C" fn foo1(x: S) -> i32 {
+        x.x
+    }
+}

--- a/tests/run-make/export/write-interface-nonexistent-dir/rmake.rs
+++ b/tests/run-make/export/write-interface-nonexistent-dir/rmake.rs
@@ -1,0 +1,18 @@
+// Regression test for <https://github.com/rust-lang/rust/issues/143981>.
+
+//@ ignore-cross-compile
+
+// NOTE: `sdylib`'s platform support is basically that of `dylib`.
+//@ needs-crate-type: dylib
+
+use run_make_support::rustc;
+
+fn main() {
+    rustc()
+        .input("libr.rs")
+        .output("does-not-exist/output")
+        .run_fail()
+        .assert_exit_code(1)
+        .assert_stderr_contains("failed to write file")
+        .assert_not_ice();
+}

